ros2topic/ros2topic/verb/hz.py

```diff
@@ -0,0 +1,337 @@
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+# This file is originally from:
+# https://github.com/ros/ros_comm/blob/6e5016f4b2266d8a60c9a1e163c4928b8fc7115e/tools/rostopic/src/rostopic/__init__.py
+
+from argparse import ArgumentTypeError
+
+import functools
+import importlib
+import math
+
+from time import sleep
+
+import rclpy
+
+from rclpy.clock import Clock
+from rclpy.clock import ClockType
+from rclpy.expand_topic_name import expand_topic_name
+from rclpy.qos import qos_profile_sensor_data
+from rclpy.validate_full_topic_name import validate_full_topic_name
+from ros2cli.node.direct import DirectNode
+from ros2topic.api import get_topic_names_and_types
+from ros2topic.api import TopicNameCompleter
+from ros2topic.verb import VerbExtension
+
+DEFAULT_WINDOW_SIZE = 10000
+
+
+def unsigned_int(string):
+    try:
+        value = int(string)
+    except ValueError:
+        value = -1
+    if value < 0:
+        raise ArgumentTypeError('value must be non-negative integer')
+    return value
+
+
+class HzVerb(VerbExtension):
+    """Print the average publishing rate to screen."""
+
+    def add_arguments(self, parser, cli_name):
+        arg = parser.add_argument(
+            'topic_name',
+            help="Name of the ROS topic to listen to (e.g. '/chatter')")
+        arg.completer = TopicNameCompleter(
+            include_hidden_topics_key='include_hidden_topics')
+        parser.add_argument(
+            '--window', '-w',
+            dest='window_size', type=unsigned_int, default=DEFAULT_WINDOW_SIZE,
+            help='window size, in # of messages, for calculating rate, '
+                 'string to (default: %d)' % DEFAULT_WINDOW_SIZE, metavar='WINDOW')
+        parser.add_argument(
+            '--filter',
+            dest='filter_expr', default=None,
+            help='only measure messages matching the specified Python expression', metavar='EXPR')
+        parser.add_argument(
+            '--wall-time',
+            dest='use_wtime', default=False, action='store_true',
+            help='calculates rate using wall time which can be helpful'
+                 ' when clock is not published during simulation')
+
+    def main(self, *, args):
+        return main(args)
+
+
+def main(args):
+    topic = args.topic_name
+    if args.filter_expr:
+        def expr_eval(expr):
+            def eval_fn(m):
+                return eval(expr)
```

Comment:
Hm.  I know this is directly ported from the original ROS1 code, but having a naked `eval` lying around here seems pretty dangerous.  In particular, the string that is passed in here can do *any* arbitrary Python thing, including copying all of your hard drive out to the network, taking command-and-control inputs from the network, or deleting your filesystem.  I don't think that sort of power is actually required here; I think a simple regular expression could probably get most of the uses.  Could you maybe explain what you think the intended use case here is?

Reply:
The filter_expr code is directly ported from the original ROS1. No changes from the first commit  ("rostopic into ros_comm" https://github.com/ros/ros_comm/commit/a82674ad87ab17e373e89d392bcf358c511026ae). One comment from code is "#694: ignore messages that don't match filter", line to https://github.com/ros/ros_comm/blob/a82674ad87ab17e373e89d392bcf358c511026ae/tools/rostopic/src/rostopic.py#L116 ", I don't know what #694 means, I guess it is internal bug track of rostopic before open source, and may noted the intended use case there. I have no idea about the requirement of the option `--filter`. Check with `echo` command of ros2topic also do not remain this option. Could we also removed here on `hz` command for code safety?

Reply:
Yeah, I did some similar code archaeology, and like you said, this option existed before history starts.  I do not know the intended use case either.  @tfoote, @dirk-thomas , do you have any further history on what the `--filter` options was meant to achieve in ROS1, and whether we still need it?  The options on the table here seem to be:

1.  Remove `--filter` entirely until a use-case comes up for it.
2.  Restrict `--filter` to be a regex only.
3.  Leave `--filter` as the current `eval`, which seems pretty dangerous.

I'd vote for 1 (we can always add it back later), but I'm interested to hear additional thoughts/opinions.